python/ray/llm/_internal/batch/stages/vllm_engine_stage.py

```diff
@@ -0,0 +1,530 @@
+"""The stage that runs vLLM engine."""
+import asyncio
+import dataclasses
+import importlib
+import logging
+import math
+import os
+import time
+import uuid
+from dataclasses import dataclass
+from pydantic import root_validator
+from typing import TYPE_CHECKING, Any, Dict, AsyncIterator, Optional, List, Tuple
+
+import ray
+from ray.llm._internal.batch.stages.base import (
+    StatefulStage,
+    StatefulStageUDF,
+)
+from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy
+
+if TYPE_CHECKING:
+    from PIL import Image
+
+logger = logging.getLogger(__name__)
+
+
+class vLLMEngineWrapper:
+    """Wrapper around the vLLM engine to handle async requests.
+
+    Args:
+        *args: The positional arguments for the engine.
+        max_pending_requests: The maximum number of pending requests in the queue.
+        runtime_env: The runtime environment to use for the vLLM engine.
+        **kwargs: The keyword arguments for the engine.
+    """
+
+    @dataclass(frozen=True)
+    class LLMRequest:
+        """A request to the LLM wrapper."""
+
+        # The request ID for the LLM engine (unique per replica).
+        request_id: int
+        # The index of the request in the batch.
+        idx_in_batch: int
+        # The full prompt string (with chat template applied if any).
+        prompt: str
+        # The images inputs for the multimodal model.
+        images: List["Image.Image"]
+        # The tokenized prompt IDs. If None, then the string prompt will be
+        # tokenized by the LLM engine. This is not recommended for performance reasons.
+        prompt_token_ids: Optional[List[int]]
+        # The sampling or pooling parameters. Use Any to avoid importing vLLM.
+        params: Any
+
+    def __init__(
+        self,
+        *args,
+        idx_in_batch_column: str,
+        max_pending_requests: int = -1,
+        runtime_env: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ):
+        self.request_id = 0
+        self.idx_in_batch_column = idx_in_batch_column
+        self.task_type = kwargs.get("task", "generate")
+
+        # Setup os.environ before importing vLLM to make sure the environment
+        # variables are effective
+        if runtime_env is not None and "env_vars" in runtime_env:
+            os.environ.update(runtime_env["env_vars"])
+
+        # Lazy import vLLM here.
+        self.vllm = importlib.import_module("vllm")
```

Comment:
This is a weird pattern. I don't like this because vllm then will be only limited to the score of this class which is why you have to make this class super long which impacts code readability, can introduce bugs, etc. You can do this on top:

```
try: 
     import vllm
except ModuleNotFoundError
     vllm = None
```

This will allow you to import this file in the context that vllm is not installed and you'd only get error when you want to really run this file at run time.

Reply:
you can put this in a utility and call it `try_import_vllm`. See these implementations in rllib for example? 

https://github.com/ray-project/ray/blob/b8def7e43e186def1f15c5aae0c7440af3dcb38d/rllib/utils/framework.py#L291

You gotta be a bit careful with type checking tho. Let's find an elegant solution for type checking?

Reply:
Yes I have no idea about how to deal with type checking in this case.